drivers: rtc: rv3028: add support for user memory access

jakubtopic opened the pull request:
This PR aims to add support for user memory access in the Micro Crystal RV3028 RTC. The device has two free RAM bytes, which are exposed as a `microcrystal,rv3028-retmem` child node, and 43 bytes of general-purpose EEPROM, accessible via `microcrystal,rv3028-eeprom` child nodes.

@bjarki-trackunit, can I ask you for feedback? :) I didn't find any driver in the tree that implements the `retained_mem` API through its children, so I am unsure if I took the right approach. ~~I am also concerned that implementing a memory clearing interface for EEPROM could shorten its lifespan.~~

jakubtopic commented on drivers/rtc/rtc_rv3028_eeprom.c:
```diff
@@ -0,0 +1,136 @@
+
+	rv3028_lock_sem(dev);
+
+	err = rv3028_enter_eerd(dev);
```
`rv3028_enter_eerd` uses busy wait. We can use sleep here

jakubtopic replied:
Updated `rv3028_eeprom_wait_busy` to use `k_msleep` if the kernel is initialized:
https://github.com/zephyrproject-rtos/zephyr/blob/186acfc1ffbc3d17182d15d3091d8e5042e785d2/drivers/rtc/rtc_rv3028.c#L206-L210

jakubtopic replied:
Updated `rv3028_eeprom_wait_busy` to use `k_msleep`
https://github.com/zephyrproject-rtos/zephyr/blob/e8bee4676e172ce1ac97965da7cc68d1e559bd34/drivers/rtc/rtc_rv3028.c#L206

nordicjm commented on drivers/rtc/rtc_rv3028_retmem.c:
```diff
@@ -0,0 +1,67 @@
+	return err;
+}
+
+const struct retained_mem_driver_api rv3028_retmem_driver_api = {
```
driver structs should be static really

jakubtopic replied:
Thanks for the feedback. Done :)

nordicjm reviewed (commented):
the driver parts should be static and in their respective directories `drivers/eeprom` and `drivers/retained_mem` really

bjarki-trackunit commented on drivers/rtc/rtc_rv3028.c:
```diff
@@ -0,0 +1,902 @@
+	return 0;
+}
+
+#ifdef CONFIG_RTC_RV3028_EEPROM
+#define RV3028_EEPROM_INIT(node, rtc_inst)                                                         \
+	static const struct rv3028_eeprom_config rv3028_eeprom_config_##node = {                   \
+		.i2c = rv3028_config_##rtc_inst.i2c,                                               \
+		.addr = DT_REG_ADDR(node),                                                         \
+		.size = DT_REG_SIZE(node),                                                         \
+	};                                                                                         \
+	DEVICE_DT_DEFINE(node, NULL, NULL, &rv3028_data_##rtc_inst, &rv3028_eeprom_config_##node,  \
+			 POST_KERNEL, CONFIG_RTC_INIT_PRIORITY, &rv3028_eeprom_driver_api);
+#else
+#define RV3028_EEPROM_INIT(node, rtc_inst)
+#endif /* CONFIG_RTC_RV3028_EEPROM */
```
You can decouple the child devices from the parent more efficiently by moving most of the code of the child devices to their own source files, including their own `DT_DRV_INST` and `DT_INST_FOREACH_STATUS_OKAY()`. Then, use `DEVICE_DT_GET(DT_INST_PARENT())` to get a pointer to the parent device, and access the shared data you need that way :)

This way, you only need to share the data struct of the parent `struct rv3028_data` and your shared functions in the header file, and keep everything else static in their own source files :)

Then, moving the source files for the child devices to their respective subsystems (eeprom and retmen) becomes trivial as well :)

bjarki-trackunit replied:
You are one of the first to do so, so absolutely understandable :)

jakubtopic committed:
dts: bindings: add binding for RV3020 user memory

Bindings for user memory access on the Micro Crystal RV3028 RTC, both for
the EEPROM child devices as well as the retained memory area.

Signed-off-by: Jakub Topic <jakub.topic@anitra.cz>

jakubtopic committed:
drivers: eeprom: add RV3028 user EEPROM access

EEPROM access is supported via the EEPROM Interface

Signed-off-by: Jakub Topic <jakub.topic@anitra.cz>

jakubtopic reviewed (commented):
I have some doubts about the need for a semaphore in the retained_mem driver.

nordicjm commented on drivers/retained_mem/retained_mem_rv3028.c:
```diff
@@ -0,0 +1,103 @@
+	const struct device *rtc_dev = config->parent;
+	int err;
+
+	rv3028_lock_sem(rtc_dev);
```
There should be mutexes here (which are only present and used if `CONFIG_RETAINED_MEM_MUTEXES` is enabled) to guard against different threads calling these functions are the same time. See implementation in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/retained_mem/retained_mem_nrf_gpregret.c for details

jakubtopic replied:
Oh, I see. I’ll remove the semaphore that guards the RTC. However, I think an additional mutex is redundant since the underlying I2C driver must prevent simultaneous transactions.

nordicjm replied:
The issue with this is that it should be usable from a bootloader which does not have multithreading, that's what the `CONFIG_RETAINED_MEM_MUTEXES` Kconfig being disabled is for, if the i2c driver forces use of a mutex/semaphore then it's out of the scope

jakubtopic replied:
I’ve removed semaphore locking from the retained memory interface implementation and am relying on the I2C API calls to be thread-safe. The dependent RTC driver still requires `k_sem`. Ensuring that the RTC driver works before the kernel is fully initialized is beyond the scope of this PR.

bjarki-trackunit commented on drivers/eeprom/Kconfig.rv3028:
```diff
@@ -0,0 +1,8 @@
+config EEPROM_RV3028
+	bool "Micro Crystal RV3028 User EEPROM support"
+	default y
+	depends on RTC_RV3028
```
You should also add this dependency
```
        depends on DT_HAS_MICROCRYSTAL_RV3028_ENABLED
```
the parent needs to have a driver, and have status "okay"

jakubtopic replied:
Hmm, I thought it is already covered by RTC_RV3028 config, which depends on it

bjarki-trackunit replied:
right, that is true, disregard my comment :)

bjarki-trackunit commented on drivers/retained_mem/Kconfig.rv3028:
```diff
@@ -0,0 +1,8 @@
+config RETAINED_MEM_RV3028
+	bool "Micro Crystal RV3028 User RAM support"
+	default y
+	depends on RTC_RV3028
```
Same here with `depends on DT_HAS_MICROCRYSTAL_RV3028_ENABLED` :)

bjarki-andreasen reviewed (dismissed):
This is great :)

jakubtopic commented:
Added Kconfig options for initialization priority of RV3028 EEPROM and retained_mem drivers, since the drivers must be initialized after the RV3028 RTC driver.

Both drivers were tested using the EEPROM/retained_mem API test suites.

jakubtopic commented on drivers/rtc/rtc_rv3028.c:
```diff
@@ -260,18 +203,22 @@ static int rv3028_eeprom_wait_busy(const struct device *dev)
 		}
 
-		k_busy_wait(RV3028_EEBUSY_POLL_US);
+		if (k_is_pre_kernel()) {
+			k_busy_wait(poll_ms * 1000);
+		} else {
+			k_msleep(poll_ms);
+		}
```
Updated `rv3028_eeprom_wait_busy` to use `k_msleep` if the kernel is initialized.

jakubtopic commented:
@nordicjm @bjarki-trackunit gentle ping

nordicjm commented on include/zephyr/drivers/rtc/rtc_rv3028.h:
```diff
@@ -0,0 +1,103 @@
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+
+void rv3028_lock_sem(const struct device *dev);
```
Really not a fan of having these as global functions, should be used with driver function pointers of the parent device in my mind. Thoughts @henrikbrixandersen ?

jakubtopic replied:
Yeah, my thoughts exactly. Forgot to ask about this.

henrikbrixandersen replied:
I agree.

jakubtopic replied:
@henrikbrixandersen any pointers how to implement it? The parent driver interface will still need to be global, right?

jakubtopic commented:
Took a step back. EEPROM/retmem is again implemented in the rtc driver and only the EEPROM/retmem interfaces are exposed in the header file. Am I missing some way to share functions between drivers without the functions being global?

Cc: @nordicjm @bjarki-trackunit

bjarki-trackunit commented:
I would just move the `rtc_rv3028.h` header to `drivers/rtc/rtc_rv3028/rtc_rv3028.h` (creating a folder for it) and include this header from the eeprom/retmem drivers manually like `zephyr_library_include_directories(${ZEPHYR_BASE}/drivers/rtc/rtc_rv3028)` and `#include "rtc_rv3028.h"`

jakubtopic commented:
Ok, done as you suggested :) Thanks, @bjarki-trackunit

bjarki-andreasen reviewed (commented):
I believe it is best to move the new drivers and the rv3028 entirely into subfolders, giving them their own sub CMakeLists.txt and Kconfig, when the addition to the "top level" Kconfig and CMakeLists.txt require more than a single entry to include the new driver :)
```
-- rtc/
  -- rtc_rv3028/
    -- rtc_rv3028.c
    -- rtc_rv3028.h
    -- Kconfig
    -- CMakeLists.txt
```
similar to how the sensor drivers do it :)

jakubtopic commented:
Alright :D I hesitated whether I should do this, because looking at the other drivers it looked like either all the drivers were in separate subfolders (like sensor) or in the same folder. But the way I did it didn't make much sense :))

Done. Thanks again :)

bjarki-trackunit commented on drivers/retained_mem/CMakeLists.txt:
```diff
@@ -5,5 +5,11 @@ zephyr_syscall_header(${ZEPHYR_BASE}/include/zephyr/drivers/retained_mem.h)
 zephyr_library()
 zephyr_library_sources_ifdef(CONFIG_USERSPACE retained_mem_handlers.c)
 zephyr_library_sources_ifdef(CONFIG_RETAINED_MEM_NRF_GPREGRET retained_mem_nrf_gpregret.c)
+zephyr_library_sources_ifdef(CONFIG_RETAINED_MEM_RV3028       retained_mem_rv3028.c)
 zephyr_library_sources_ifdef(CONFIG_RETAINED_MEM_ZEPHYR_RAM   retained_mem_zephyr_ram.c)
 zephyr_library_sources_ifdef(CONFIG_RETAINED_MEM_ZEPHYR_REG   retained_mem_zephyr_reg.c)
+zephyr_library_sources_ifdef(CONFIG_RETAINED_MEM_RV3028       retained_mem_rv3028.c)
```
It seems this file is included twice?

jakubtopic replied:
I remember noticing that, but I forgot to fix it :))

bjarki-andreasen reviewed (commented):
Could you move the eeprom and retmem drivers into their own folders as well, like the rtc_rv3028? :)

bjarki-andreasen reviewed (dismissed):
Really nice work :)

bjarki-andreasen commented:
Ping @nordicjm

henrikbrixandersen reviewed (changes requested):
Blocking until I have had time to review.

fabiobaltieri commented on drivers/eeprom/eeprom_rv3028/eeprom_rv3028.c:
```diff
@@ -0,0 +1,179 @@
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, rv3028_eeprom_init, NULL, NULL, &rv3028_eeprom_config_##inst,  \
+			      POST_KERNEL, CONFIG_EEPROM_RV3028_INIT_PRIORITY,                     \
```
temping to suggest making this `CONFIG_RTC_INIT_PRIORITY`, the option is going to be available because you depend on the driver that depends on RTC, if you initialize both the parent and child at the same priority then the subpriority stuff will ensure that they are initialized in the correct sequence and you'd save one Kconfig option to mess with.

Con is that it looks odd to have an eeprom device with an rtc priority, but then it already a custom symbol so why not.

What do you think?

jakubtopic replied:
Sounds reasonable. Thanks for the suggestion.

fabiobaltieri commented on drivers/retained_mem/retained_mem_rv3028/retained_mem_rv3028.c:
```diff
@@ -0,0 +1,86 @@
+
+#define RV3028_USER_RAM_SIZE (2 * sizeof(uint8_t))
+
+BUILD_ASSERT(CONFIG_RETAINED_MEM_RV3028_INIT_PRIORITY > CONFIG_RTC_INIT_PRIORITY,
+	     "RV3028 Retained Memory driver must be initialized after RV3028 RTC driver");
```
same comments for the priority

fabiobaltieri commented on drivers/retained_mem/retained_mem_rv3028/retained_mem_rv3028.c:
```diff
@@ -0,0 +1,86 @@
+{
+	ARG_UNUSED(dev);
+
+	return (ssize_t)RV3028_USER_RAM_SIZE;
```
drop the casting, does not do anything

jakubtopic commented on drivers/eeprom/eeprom_rv3028/eeprom_rv3028.c:
```diff
@@ -0,0 +1,179 @@
+#define RV3028_USER_EEPROM_SIZE 0x2B
+
+BUILD_ASSERT(CONFIG_EEPROM_RV3028_INIT_PRIORITY > CONFIG_RTC_INIT_PRIORITY,
+	     "RV3028 EEPROM driver must be initialized after RV3028 RTC driver");
```
this validation is indeed redundant. I've removed it along with the custom symbol

nordicjm commented on drivers/retained_mem/CMakeLists.txt:
```diff
@@ -3,7 +3,10 @@
 zephyr_library_sources_ifdef(CONFIG_RETAINED_MEM_ZEPHYR_RAM   retained_mem_zephyr_ram.c)
 zephyr_library_sources_ifdef(CONFIG_RETAINED_MEM_ZEPHYR_REG   retained_mem_zephyr_reg.c)
+
+add_subdirectory_ifdef(CONFIG_RETAINED_MEM_RV3028 retained_mem_rv3028)
```
we do not use sub-folders for drivers

bjarki-andreasen replied:
I suggested using subfolders here as these drivers don't occupying just a single line of the `CMakeLists.txt`. The drivers include private headers like `${ZEPHYR_BASE}/drivers/rtc/rtc_rv3028` which should not be leaked to other RTC drivers, hence the:

```
zephyr_library()
zephyr_library_sources(retained_mem_rv3028.c)
zephyr_library_include_directories(${ZEPHYR_BASE}/drivers/rtc/rtc_rv3028)
```

and adding multiple lines to the "parent" CMakeLists.txt per driver is also going to be a mess quickly. It is harmless and IMO cleaner.

nordicjm replied:
We have drivers adding multiple lines to cmake files, would follow what we already have e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/CMakeLists.txt

jakubtopic replied:
Without subfolders, we cannot construct a new target using only the `zephyr_library_*` functions. As @bjarki-andreasen mentioned, the private headers of the rtc_rv3028 driver would be included in all retained_mem/eeprom drivers.

nordicjm replied:
```
if(SYMBOL_NAME)
  ...code here...
endif()
```

jakubtopic replied:
I understand that Zephyr’s review model is not well supported by GitHub, making it hard to track follow-up changes. However, this was one of my previous solutions. The issue with this approach is that when the rv3028 retmem/eeprom driver is enabled, the rv3028 headers become accessible to any other retmem/eeprom drivers that are also enabled.

By having CMakeLists in a separate subfolder, we can construct a new library target that keeps the rv3028 headers more private.

Another solution would be to use plain CMake functions instead of Zephyr’s wrapper interface, but that’s outside the scope of this PR.

henrikbrixandersen replied:
> The issue with this approach is that when the rv3028 retmem/eeprom driver is enabled, the rv3028 headers become accessible to any other retmem/eeprom drivers that are also enabled.

That's not really an issue.

jakubtopic replied:
Ok, maybe I took it too far :) I'll revert it back to a "flat" structure

nordicjm replied:
Put that one private file under a different name e.g. rtc_rv_3028_priv.h ? I don't see the issue. There are private files in subsystems in zephyr that you can't just include by doing `#include <system/priv_file.h>` but if someone really wants to include it they can do `#include <../../../system/priv_file.h>`

nordicjm replied:
Also mark the API as hidden using doxygen comments

nordicjm commented on drivers/retained_mem/Kconfig:
```diff
@@ -34,7 +34,8 @@ module-str = retained_mem
 source "subsys/logging/Kconfig.template.log_config"
 
 source "drivers/retained_mem/Kconfig.nrf"
-
```
?

nordicjm commented on dts/bindings/retained_mem/microcrystal,rv3028-retmem.yaml:
```diff
@@ -0,0 +1,26 @@
+          reg = <0x52>;
+          backup-switch-mode = "disabled";
+
+          rtc_retmem: rv3028-retmem {
+                  compatible: "microcrystal,rv3028-retmem";
+                  status = "okay";
+          };
```
this should, like the eeprom above, have a reg for offset and size

jakubtopic replied:
Originally, I didn’t think it would be useful when there were only two bytes available to use. If I were to add the reg property, I’d like some advice on what base address to use specifically. The EEPROM and RAM in the RTC have separate address spaces, and I want to avoid collisions between the unit addresses of the child nodes. User EEPROM occupies addresses 0x00 to 0x2A, while User RAM has addresses 0x1F and 0x20.

nordicjm replied:
you use the address space that each occupies, so 0x1f to 0x20, they are separate nodes

jakubtopic replied:
yes but the child nodes then might have the same unit address in devicetree. however since the unit addresses are optional, this might be a non-issue...

bjarki-andreasen replied:
Looking at the memory layout, the EEPROM is in a different address space than RAM, I would use the address of the control registers in RAM 0x30 to 0x37 for the EEPROM node (since this is what the driver interacts with):
```
          rtc_eeprom: rv3028-eeprom@30 {
                  compatible: "microcrystal,rv3028-eeprom";
                  status = "okay";
                  reg = <0x30 0x7>;
          };
```

jakubtopic replied:
> I would use the address of the control registers in RAM 0x30 to 0x37 for the EEPROM node

Hmm... this would prevent configuring the area used in eeprom

jakubtopic replied:
Oh, you're right :) should the value of the reg property be constant in the binding?

I'll add a sample.

jakubtopic replied:
@bjarki-andreasen, I followed @nordicjm's suggestion and used the `reg` property to define offset and size for both nodes. It works, dts scripts do not complain, but it still doesn't feel right to me. Neither EEPROM nor RAM addresses belong in the parent node's address space. Wouldn't it be better to use some custom properties that are not mapped to unit addresses?

Here's how the current config looks:
```
&i2c {
  ext_rtc: rv3028@52 {
      compatible = "microcrystal,rv3028";
      reg = <0x52>;
      backup-switch-mode = "disabled";
      status = "okay";

      #address-cells = <1>;
      #size-cells = <1>;

      rtc_eeprom: rv3028-eeprom@0 {
          compatible = "microcrystal,rv3028-eeprom";
          reg = <0x0 0x2b>;
          status = "okay";
      };

      rtc_retmem: rv3028-retmem@1f {
          compatible = "microcrystal,rv3028-retmem";
          reg = <0x1f 0x2>;
          status = "okay";
      };
  };
};
```

jakubtopic replied:
@decsny I'd appreciate your thoughts on this

decsny replied:
can you clarify what is your concern?

nordicjm commented on drivers/retained_mem/retained_mem_rv3028/retained_mem_rv3028.c:
```diff
@@ -0,0 +1,83 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(rv3028, CONFIG_RETAINED_MEM_LOG_LEVEL);
+
+#define RV3028_USER_RAM_SIZE (2 * sizeof(uint8_t))
```
then this would be dropped, could have a assert to ensure it offset is not less than the first RAM address and offset + length are not above the second RAM address

nordicjm commented on drivers/retained_mem/retained_mem_rv3028/retained_mem_rv3028.c:
```diff
@@ -0,0 +1,83 @@
+	const struct rv3028_retained_mem_config *config = dev->config;
+	const struct device *rtc_dev = config->parent;
+
+	return rv3028_read_regs(rtc_dev, (RV3028_REG_USER_RAM1 + offset), buffer, size);
+}
+
+static int rv3028_retained_mem_write(const struct device *dev, off_t offset, const uint8_t *buffer,
+				     size_t size)
+{
+	const struct rv3028_retained_mem_config *config = dev->config;
+	const struct device *rtc_dev = config->parent;
+
+	return rv3028_write_regs(rtc_dev, (RV3028_REG_USER_RAM1 + offset), buffer, size);
```
These functions would then use the reg, someone might not want an application to look or update the first byte and only use the second, etc.

nordicjm commented on drivers/retained_mem/retained_mem_rv3028/retained_mem_rv3028.c:
```diff
@@ -0,0 +1,83 @@
+	const struct device *rtc_dev = config->parent;
+	uint8_t block[RV3028_USER_RAM_SIZE] = {0};
+
+	return rv3028_write_regs(rtc_dev, RV3028_REG_USER_RAM1, block, RV3028_USER_RAM_SIZE);
```
likewise this should only clear what is configured

nordicjm commented on drivers/retained_mem/retained_mem_rv3028/retained_mem_rv3028.c:
```diff
@@ -0,0 +1,83 @@
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, rv3028_retmem_init, NULL, NULL,                                \
+			      &rv3028_retained_mem_config_##inst, POST_KERNEL,                     \
+			      CONFIG_RTC_INIT_PRIORITY, &rv3028_retmem_api);
```
`CONFIG_RETAINED_MEM_INIT_PRIORITY`

jakubtopic replied:
Using the parent (RTC) init priority was cautiously suggested by @fabiobaltieri, and I like this approach despite it potentially looking odd. The RV3028 is an I2C RTC, hence it is initialized relatively late by default. I don’t want to require users to push the retmem init priority past I2C and RTC, as there could be other retmem devices that do not depend on other components.

nordicjm replied:
this is a retained memory driver, it needs to use the retention init priority. Code (application and driver) should know when things init, it makes no sense and only provides mass confusion if someone wants to use this but needs to use it earlier if they reduce the retained memory init priority, use it, and just undefined operation and then have to spend a long time trying to work out what's going wrong when the issue is caused by a driver class using what is basically a completely random init priority that it should not be using. The nodes should be init'd in order, there is a python script which will check the priorities so if someone sets retained memory lower than the rtc the script should throw an error (can try that in a sample build and see)

jakubtopic replied:
it's not a random init priority. it's an init priority of the actual device driver. this is not a standalone retmem device. this is an RTC implementing a retmem interface.

if this causes confusion as you suggest, I would revert to the previous solution - a custom init priority symbol (`CONFIG_RETAINED_MEM_RV3028_INIT_PRIORITY`), that would be visible from menuconfig

jakubtopic replied:
Reverted to the custom init symbols

jakubtopic committed:
drivers: retained_mem: add RV3028 user memory access

User RAM access is supported using the Retained Memory API

Signed-off-by: Jakub Topic <jakub.topic@anitra.cz>

jakubtopic commented:
- reverted EEPROM/retmem driver init priority to custom symbols to avoid confusion
- moved the drivers back from subfolders
- added reg property for retmem offset and size
- rebased to main

nordicjm commented on drivers/rtc/rtc_rv3028.h:
```diff
@@ -0,0 +1,92 @@
+#include <zephyr/device.h>
+#include <sys/types.h>
+
+void rv3028_lock_sem(const struct device *dev);
+
+void rv3028_unlock_sem(const struct device *dev);
+
+int rv3028_read_regs(const struct device *dev, uint8_t addr, void *buf, size_t len);
+
+int rv3028_read_reg8(const struct device *dev, uint8_t addr, uint8_t *val);
+
+int rv3028_write_regs(const struct device *dev, uint8_t addr, const void *buf, size_t len);
+
+int rv3028_write_reg8(const struct device *dev, uint8_t addr, uint8_t val);
+
+int rv3028_update_reg8(const struct device *dev, uint8_t addr, uint8_t mask, uint8_t val);
+
+int rv3028_eeprom_wait_busy(const struct device *dev, int poll_ms);
+
+int rv3028_exit_eerd(const struct device *dev);
+
+int rv3028_enter_eerd(const struct device *dev);
+
+int rv3028_eeprom_command(const struct device *dev, uint8_t command);
```
bit confused by this, didn't we agree to do this with function pointers?

bjarki-andreasen replied:
could you provide an example of what you mean?

nordicjm replied:
like how any driver works, sub-drivers should get a function pointer from the parent device if they need to access these functions and in a private header. That's what was agreed https://github.com/zephyrproject-rtos/zephyr/pull/74034#discussion_r1639551300

bjarki-andreasen replied:
I meant a code example... For a device driver specific header, there is no need for function pointers, the functions are not generic/interchangeable, so why add the complexity of a table of function pointers? (again, guessing what you mean by function pointers...)

nordicjm replied:
Because these are private functions, they should not be user-callable, they should only be implemented by the base driver and called by the other driver classes using this base driver

jakubtopic replied:
I would really appreciate a code example of this. I'm not exactly sure how to declare a function pointer table without making it callable by the user.

bjarki-andreasen replied:
Regarding "user callable" this header is private, its in `drivers/rtc`, and included like `#include "rtc_rv3028"`. Its not in `include/zephyr/drivers/rtc`, it has to be included manually through CMake, hence `zephyr_library_include_directories(${ZEPHYR_BASE}/drivers/rtc)`

nordicjm replied:
I missed that it was in that folder and not in includes, solution is acceptable

nordicjm reviewed (dismissed):
Retained memory part OK

henrikbrixandersen commented:
@jakubtopic Please rebase on `main` and resolve any merge conflicts.

bjarki-andreasen commented:
@jakubtopic could you fix the merge conflicts?

jakubtopic commented:
Rebased to main

bjarki-andreasen commented:
ping @henrikbrixandersen @nordicjm

henrikbrixandersen commented on drivers/eeprom/CMakeLists.txt:
```diff
@@ -26,3 +26,8 @@ zephyr_library_sources_ifdef(CONFIG_EEPROM_FAKE eeprom_fake.c)
+
+if(CONFIG_EEPROM_RV3028)
+  zephyr_library_sources(eeprom_rv3028.c)
+  zephyr_library_include_directories(${ZEPHYR_BASE}/drivers/rtc)
```
Sorry, just saw this now. This is not the best way to do it, I'm afraid. It's true that we have some legacy implementations in tree doing something similar, but for this new functionality I'd suggest implementing a base MFD device driver, layering both the RTC and EEPROM drivers on top of that. This avoids these somewhat weird dependencies between an EEPROM driver and a RTC driver.

bjarki-andreasen replied:
so `drivers/mfd/mfd_rv3028.c` and

```
if(CONFIG_EEPROM_RV3028)
  zephyr_library_sources(eeprom_rv3028.c)
  zephyr_library_include_directories(${ZEPHYR_BASE}/drivers/mfd)
  ...
```

henrikbrixandersen replied:
No. The header file goes under `include/zephyr/drivers/mfd/`.

henrikbrixandersen replied:
Private how? More private than the other MFD driver APIs in the three, which are placed in the include folder?

Specifying the MFD driver API in a public header file allows for easier out-of-tree usage and prevents odd dependencies between otherwise unrelated driver subsystems.

bjarki-andreasen replied:
I do agree, but was this not discussed earlier https://github.com/zephyrproject-rtos/zephyr/pull/74034#discussion_r1639557003 where it seemed we wanted to keep the header private? @nordicjm?

bjarki-andreasen replied:
ok, in that case, the `drivers/rtc/rtc_rv3028.h` header does seem fine if the names are updated to use the `mfd_` namespace, and can be moved to `zephyr/drivers/mfd` I believe, maybe with the name of the lock/unlock functions updated to not include the `_sem_` part? locking is not a private implementation detail IMO :)

There is precedent for something a bit wilder in there:
https://github.com/zephyrproject-rtos/zephyr/blob/318b49570a77ce631dc3d97ad7e7a05df5459d8e/include/zephyr/drivers/mfd/nct38xx.h#L25
and
https://github.com/zephyrproject-rtos/zephyr/blob/318b49570a77ce631dc3d97ad7e7a05df5459d8e/include/zephyr/drivers/mfd/adp5585.h#L102-L104

we may want to start setting and documenting some rules for mfd drivers :)

henrikbrixandersen replied:
Sounds good to me.

bjarki-andreasen commented:
To summarize, hopefully correctly @jakubtopic and  @henrikbrixandersen:

* Rename `rtc_rv3028.h` -> `mfd_rv3028.h`
* Namespace functions in mfd_rv3028.h to `mdf_`
  `rv3028_read_regs` -> `mfd_rv3028_read_regs`
* Move `mfd_rv3028.h` to `include/zephyr/drivers/mfd/`
* Convert for both eeprom and retained_mem partial drivers:
  ```
  if(CONFIG_EEPROM_RV3028)
  zephyr_library_sources(eeprom_rv3028.c)
  zephyr_library_include_directories(${ZEPHYR_BASE}/drivers/rtc)
  ```
  to
  ```
  zephyr_library_sources_ifdef(CONFIG_EEPROM_RV3028 eeprom_rv3028.c)
  ```
* Update drivers to use the updated names and include path of `mfd_rv3028.h`
  `#include "rtc_rv3028.h"` -> `#include <zephyr/drivers/mfd/mfd_rv3028.h>`

henrikbrixandersen commented:
> To summarize, hopefully correctly @jakubtopic and @henrikbrixandersen:

Won't the existing `drivers/rtc/rtc_rv3028.c` need to be converted to use the new drivers/mfd/mfd_rv3028.c` driver API as well?

bjarki-andreasen commented:
> > To summarize, hopefully correctly @jakubtopic and @henrikbrixandersen:
> 
> Won't the existing `drivers/rtc/rtc_rv3028.c` need to be converted to use the new drivers/mfd/mfd_rv3028.c` driver API as well?

Right, forgot about that :)

henrikbrixandersen commented:
Please rebase on `main`, resolve any merge conflicts, and address the review comments.

nordicjm commented:
@jakubtopic are you coming back to this?
